### PR TITLE
Null safety

### LINKED ISSUE #85: migrate to null-saftey
- https://dart.dev/null-safety/understanding-null-safety
-  https://dart.dev/null-safety/migration-guide
- https://dart.dev/null-safety/unsound-null-safety

### PR DESCRIPTION
**Pull request by WieFel**
Closes #85

### CODE REVIEW

File: lib/scaffold.dart [hunk 1]
```diff
@@ -571,8 +571,11 @@ class BackdropScaffoldState extends State<BackdropScaffold>
               children: <Widget>[
                 Flexible(
                   child: _MeasureSize(
-                    onChange: (size) =>
-                        setState(() => _backPanelHeight = size.height),
+                    onChange: (size) {
+                      if (size != null) {
+                        setState(() => _backPanelHeight = size.height);
+                      }
+                    },
```
Comment (WieFel) [inline lib/scaffold.dart:578]: Added enclosing `size != null` check. Is this ok?

File: lib/scaffold.dart [hunk 2]
```diff
@@ -600,10 +603,13 @@ class BackdropScaffoldState extends State<BackdropScaffold>
               children: <Widget>[
                 // subHeader
                 _MeasureSize(
-                  onChange: (size) =>
-                      setState(() => _subHeaderHeight = size.height),
+                  onChange: (size) {
+                    if (size != null) {
+                      setState(() => _subHeaderHeight = size.height);
+                    }
+                  },
```
Comment (WieFel) [inline lib/scaffold.dart:610]: Added enclosing `size != null` check. Is this ok?

File: lib/scaffold.dart [hunk 3]
```diff
@@ -618,7 +624,7 @@ class BackdropScaffoldState extends State<BackdropScaffold>
     );
   }
 
-  Future<bool> _willPopCallback(BuildContext context) async {
+  Future<bool?> _willPopCallback(BuildContext context) async {
     if (isBackLayerRevealed) {
       concealBackLayer();
       return null;
```
Comment (WieFel) [inline lib/scaffold.dart:630]: Could `_willPopCallback` also return something non-null?

File: lib/scaffold.dart [hunk 4]
```diff
@@ -633,7 +639,8 @@ class BackdropScaffoldState extends State<BackdropScaffold>
 
   Widget _buildBody(BuildContext context) {
     return WillPopScope(
-      onWillPop: () => _willPopCallback(context),
+      onWillPop: (() => _willPopCallback(context) as Future<bool>)
+          as Future<bool> Function()?,
```
Comment (WieFel) [inline lib/scaffold.dart:643]: Can be changed if we are able to change return type of `_willPopCallback` to `Future<bool>`.

File: lib/scaffold.dart [hunk 5]
```diff
@@ -707,12 +714,12 @@ class BackdropScaffoldState extends State<BackdropScaffold>
 /// Credit: https://stackoverflow.com/a/60868972/2554745
 class _MeasureSize extends StatefulWidget {
   final Widget child;
-  final ValueChanged<Size> onChange;
+  final ValueChanged<Size?> onChange;
```
Comment (WieFel) [inline lib/scaffold.dart:717]: Can't do anything to change `Size?` to `Size` here, right?

**Review by WieFel — COMMENTED** [review]
@daadu need your help on some things

**Comment by WieFel** [comment]
Also: should `example/` and `demo/` also be migrated?

For `demo` we need `gallerize` package to be migrated, which still waits on flutter_highlight (https://github.com/git-touch/highlight/issues/18) to be migrated.

File: lib/scaffold.dart (hunk 5 above)
Comment (daadu) [reply lib/scaffold.dart]: Revert back to `ValueChanged<Size>` in https://github.com/fluttercommunity/backdrop/pull/86/commits/2f4010e5619ad2089290835c3e40b7048594943f.

I have not studied "null safety" completely yet but added a null check before calling the "callback" should work.

Resolve this if you think that would work.

File: lib/scaffold.dart (hunk 4 above)
Comment (daadu) [reply lib/scaffold.dart:637]: Resolved in https://github.com/fluttercommunity/backdrop/commit/2f4010e5619ad2089290835c3e40b7048594943f.

File: lib/scaffold.dart (hunk 1 above)
Comment (daadu) [reply lib/scaffold.dart]: Resolved in https://github.com/fluttercommunity/backdrop/commit/2f4010e5619ad2089290835c3e40b7048594943f.

**Comment by daadu** [comment]
@WieFel Should we merge this to master?

 Or release "null-saftey" to different version series "v0.6.x". This is until we are supporting "non-null safety" in v0.5.x.

Any new fix, will need to be merged with both. Later we can merge them.

**Comment by daadu** [comment]
> @WieFel Should we merge this to master?
> 
> Or release "null-saftey" to different version series "v0.6.x". This is until we are supporting "non-null safety" in v0.5.x.
> 
> Any new fix, will need to be merged with both. Later we can merge them.

@WieFel Give me your views on it.

If the user does not uprgade his project to "null safety", will our package break things? if that is true, then we can merge this to master.

File: lib/scaffold.dart (hunk 2 above)
Comment (daadu) [reply lib/scaffold.dart]: Resolved in https://github.com/fluttercommunity/backdrop/commit/2f4010e5619ad2089290835c3e40b7048594943f.

File: lib/scaffold.dart (hunk 3 above)
Comment (daadu) [reply lib/scaffold.dart]: Resolved in https://github.com/fluttercommunity/backdrop/pull/86/commits/c77c561f2f754576242980531f8e08710632e2a0

File: lib/scaffold.dart (hunk 4 above)
Comment (daadu) [reply lib/scaffold.dart]: Yeah, not returning "null" in `_willPopCallback`, makes sense.

I have pushed the changes in https://github.com/fluttercommunity/backdrop/pull/86/commits/c77c561f2f754576242980531f8e08710632e2a0.

Resolve this, if you think they are ok?

**Comment by WieFel** [comment]
> If the user does not uprgade his project to "null safety", will our package break things? if that is true, then we can merge this to master.

I just played around with it.  I downgraded Flutter to version `1.22.6` and created a sample app into which I imported backdrop `null-safety` branch. When doing `flutter pub get`, the following error arises:
```
The current Dart SDK version is 2.10.5.

Because backdrop_nullsafety_test depends on backdrop from git which requires SDK version >=2.12.0 <3.0.0, version solving failed.
pub get failed (1; Because backdrop_nullsafety_test depends on backdrop from git which requires SDK version >=2.12.0 <3.0.0, version solving failed.)
```
(where `backdrop_nullsafety_test` is the name of my freshly created app)

Thus, by applying null-safe code to master we would force users to upgrade to Flutter `2.0.0`, which isn't so bad in my opinion.

**Comment by proninyaroslav** [comment]
@WieFel 
The library is still actively developing and I don't think that such a strict separation is necessary, thereby creating legacy at an early stage. API essentially remains the same. Switching to the null safety version is not painful when using `backdrop` in a project.

**Comment by daadu** [comment]
@WieFel @proninyaroslav Ok then, lets release this as `v0.6.0`, with a note in README that explains the breaking change.

@WieFel Can you add that note?

**Comment by daadu** [comment]
> @WieFel Can you add that note?


Added myself.

**Review by daadu — APPROVED** [review]

**Comment by WieFel** [comment]
LGTM 🎉

**Comment by daadu** [comment]
This fix/feature has landed with [v0.6.0 on pub.dev](https://pub.dev/packages/backdrop/versions/0.6.0).

**Pull request merged**